### PR TITLE
intel/openfl#429 Add `tf.data` usage example with new tutorial

### LINKED ISSUE #429: Create example for tf.data with the ShardDescriptor interface

### PR DESCRIPTION
**Pull request by MasterSkepticista**

### CODE REVIEW

**Comment by MasterSkepticista** [comment]
***I have read the CLA Document and I hereby sign the CLA***

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/director/start_director_with_tls.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/director/start_director_with_tls.sh]: Since the README describes running the experiment without TLS, we can get rid of this file.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+from typing import List, Tuple
+
+import tensorflow as tf
+
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+logger = logging.getLogger(__name__)
+
+
+class CIFAR10ShardDescriptor(ShardDescriptor):
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:15]: Could we split up this class into ShardDescriptor and ShardDataset?

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+params:
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml:1]: We can get rid of this file since README describes copying this from the first envoy.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/start_envoy_with_tls.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/start_envoy_with_tls.sh]: Let's remove this file.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+   "outputs": [],
+   "source": [
+    "import time\n",
+    "from tensorflow.keras.utils import Progbar\n",
+    "\n",
+    "TI = TaskInterface()\n",
+    "\n",
+    "@TI.register_fl_task(model='model', data_loader='dataset', optimizer='optimizer', device='device')     \n",
+    "def train(model, dataset, optimizer, device, loss_fn=loss_fn, warmup=False):\n",
+    "    start_time = time.time()\n",
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb]: `start_time` isn't being used.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.8.10"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Comment (mansishr) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb:375]: Please clear metadata from the notebook

**Review by mansishr — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/director/start_director_with_tls.sh (hunk 1 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/director/start_director_with_tls.sh]: Noted.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py (hunk 2 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:15]: Good point. This was in fact a primary hindrance to directly using `tf.data` API. [`tf.data.Dataset`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset) returns a composite object with extensions like batching, shuffling, etc. 

Because it is an iterator itself by design, it eliminates the need for a `ShardDataset` object (which from the interface definition is simply an iterator over `x` and `y` with `__getitem__`)

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml (hunk 3 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml:1]: This was cloned directly from the `Tensorflow_MNIST` example. I will take a look.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb (hunk 6 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb:375]: Metadata seems to be empty in each cell (and so the output). I believe the other stuff remains as-is. Could you re-check?

I attempted clearing all cell outputs, but this last one remains for some reason.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+        test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+        return train_ds, test_ds
+
+    def get_shard_dataset_types(self) -> List[str]:
+        """Get available split names"""
+        return list(self.splits)
+
+    def get_split(self, name: str) -> tf.data.Dataset:
+        """Return a shard dataset by type."""
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:68]: Wrong description. Looks like this is for previous function

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+        # Split
+        x_train, y_train = x_train[rank-1::worldsize], y_train[rank-1::worldsize]
+        x_test, y_test = x_test[rank-1::worldsize], y_test[rank-1::worldsize]
+
+        train_ds = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+        test_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test))
+        return train_ds, test_ds
+
+    def get_shard_dataset_types(self) -> List[str]:
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:63]: Wrong description

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
+# Copyright (C) 2020-2021 Intel Corporation
```
Comment (shreyasnc1) [inline openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py]: 2022

**Review by shreyasnc1 — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py (hunk 8 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:63]: Looks fair to me. This method returns the available `split` names (list of strings)

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py (hunk 7 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:68]: Returns the dataset as per the split name specified. Not for the previous function. Do you mean adding more docstrings?

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py (hunk 9 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py]: Fixed in the last week commit ca468dd

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb (hunk 6 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/workspace/Tensorflow_CIFAR.ipynb:375]: Should be OK I believe. The last cell metadata is non-identifiable and wasn't cleared even upon clearing cell outputs. Marking it as resolved.

Open to alternative ways.

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml (hunk 3 above)
Comment (MasterSkepticista) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/envoy_config_two.yaml:1]: 3c94773

File: openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py (hunk 2 above)
Comment (ladi-pomsar) [reply openfl-tutorials/interactive_api/Tensorflow_CIFAR_tfdata/envoy/cifar10_shard_descriptor.py:15]: I think this is really a question of whether you want to adhere to the ShardDescriptor and ShardDataset design. IMHO, it doesn't make sense in this case, unless you want to enforce it strictly.

**Comment by MasterSkepticista** [comment]
Is this OK to merge?

**Review by psfoley — APPROVED** [review]
LGTM!
